File: same tree.py
# using preorder treversal
# time: O(n) and space: O(1)
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if p == None and q == None:
            return True
        if p == None or q == None:
            return False
        if p.val != q.val:
            return False
        left = self.isSameTree(p.left, q.left)
        right = self.isSameTree(p.right, q.right)
        return left and right

File: same tree.cpp
// time: O(n) and space: O(n) [ stack space]
class Solution {
public:
    bool solve(TreeNode * p, TreeNode * q){
        if (p == nullptr && q == nullptr){
            return true;
        }
        if (p == nullptr || q == nullptr){
            return false;
        }
        if (p->val != q->val) return false;
        bool left = solve(p->left, q->left);
        bool right = solve(p->right, q->right);
        return left && right;
    }
    bool isSameTree(TreeNode* p, TreeNode* q) {
        return solve(p, q);
    }
};
